src/OpenGL/Extensions/Silk.NET.OpenGL.Extensions.ImGui/ImGuiController.cs

```diff
@@ -234,211 +146,390 @@ public void Update(float deltaSeconds)
-            io.KeyMap[(int)ImGuiKey.Backspace] = (int)Key.Backspace;
-            io.KeyMap[(int)ImGuiKey.Enter] = (int)Key.Enter;
-            io.KeyMap[(int)ImGuiKey.Escape] = (int)Key.Escape;
-            io.KeyMap[(int)ImGuiKey.A] = (int)Key.A;
-            io.KeyMap[(int)ImGuiKey.C] = (int)Key.C;
-            io.KeyMap[(int)ImGuiKey.V] = (int)Key.V;
-            io.KeyMap[(int)ImGuiKey.X] = (int)Key.X;
-            io.KeyMap[(int)ImGuiKey.Y] = (int)Key.Y;
-            io.KeyMap[(int)ImGuiKey.Z] = (int)Key.Z;
+            var io = ImGuiNET.ImGui.GetIO();
+            io.KeyMap[(int) ImGuiKey.Tab] = (int) Key.Tab;
+            io.KeyMap[(int) ImGuiKey.LeftArrow] = (int) Key.Left;
+            io.KeyMap[(int) ImGuiKey.RightArrow] = (int) Key.Right;
+            io.KeyMap[(int) ImGuiKey.UpArrow] = (int) Key.Up;
+            io.KeyMap[(int) ImGuiKey.DownArrow] = (int) Key.Down;
+            io.KeyMap[(int) ImGuiKey.PageUp] = (int) Key.PageUp;
+            io.KeyMap[(int) ImGuiKey.PageDown] = (int) Key.PageDown;
+            io.KeyMap[(int) ImGuiKey.Home] = (int) Key.Home;
+            io.KeyMap[(int) ImGuiKey.End] = (int) Key.End;
+            io.KeyMap[(int) ImGuiKey.Delete] = (int) Key.Delete;
+            io.KeyMap[(int) ImGuiKey.Backspace] = (int) Key.Backspace;
+            io.KeyMap[(int) ImGuiKey.Enter] = (int) Key.Enter;
+            io.KeyMap[(int) ImGuiKey.Escape] = (int) Key.Escape;
+            io.KeyMap[(int) ImGuiKey.A] = (int) Key.A;
+            io.KeyMap[(int) ImGuiKey.C] = (int) Key.C;
+            io.KeyMap[(int) ImGuiKey.V] = (int) Key.V;
+            io.KeyMap[(int) ImGuiKey.X] = (int) Key.X;
+            io.KeyMap[(int) ImGuiKey.Y] = (int) Key.Y;
+            io.KeyMap[(int) ImGuiKey.Z] = (int) Key.Z;
         }
 
-        private unsafe void RenderImDrawData(ImDrawDataPtr drawData)
+        private unsafe void SetupRenderState(ImDrawDataPtr drawDataPtr, int fbWidth, int fbHeight)
         {
-            uint vertexOffsetInVertices = 0;
-            uint indexOffsetInElements = 0;
+            // Setup render state: alpha-blending enabled, no face culling, no depth testing, scissor enabled, polygon fill
+            _gl.Enable(GLEnum.Blend);
+            _gl.BlendEquation(GLEnum.FuncAdd);
+            _gl.BlendFuncSeparate(GLEnum.SrcAlpha, GLEnum.OneMinusSrcAlpha, GLEnum.One, GLEnum.OneMinusSrcAlpha);
+            _gl.Disable(GLEnum.CullFace);
+            _gl.Disable(GLEnum.DepthTest);
+            _gl.Disable(GLEnum.StencilTest);
+            _gl.Enable(GLEnum.ScissorTest);
+            _gl.Disable(GLEnum.PrimitiveRestart);
+            _gl.PolygonMode(GLEnum.FrontAndBack, GLEnum.Fill);
+
+            // Setup viewport, orthographic projection matrix
+            // Our visible imgui space lies from draw_data->DisplayPos (top left) to draw_data->DisplayPos+data_data->DisplaySize (bottom right). DisplayPos is (0,0) for single viewport apps.
+            _gl.Viewport(0, 0, (uint) fbWidth, (uint) fbHeight);
```

Comment:
In the old implementation the user was responsible for settings this (see the `Resize` event binding in the `ImGui` lab application you added)

Can you try removing these lines to better fit the old implementation? Otherwise someone, somewhere will be doing strange viewport things and run into a nasty bug...

Reply:
So I tried to comment his out, but when I move the window to a retina display and then resize, it gets all messed up.

The reference design that we implemented here is storing the original viewport and then putting it back:

https://github.com/jsm174/Silk.NET/commit/4a7e4e16543df16af82d6124c64070d4d4987796#diff-d52baa3bc86c4692b0e9cb46dcddda2e62efc392d321e88ef75338260ecabb01L440

<img width="696" alt="Screen Shot 2021-04-21 at 5 35 08 PM" src="https://user-images.githubusercontent.com/1197137/115624035-0a021680-a2c8-11eb-97fe-ccf2c90f9ef9.png">
<img width="1155" alt="Screen Shot 2021-04-21 at 5 35 17 PM" src="https://user-images.githubusercontent.com/1197137/115624029-09698000-a2c8-11eb-823d-0f009e56679c.png">

Reply:
This is probably a problem with the sample more than anything else as we’re using Resize instead of FramebufferResize, so after fixing that removing these lines should still be ok.

Reply:
Fixed in force push to the third commit.